File: runtime/fx.lua
-- Moonshine-style post FX + a Shadertoy → LÖVE converter.
-- Seek-safe: shaders + uniforms only. No clocks. Applied to an offscreen
-- canvas of an `s:fx` group.
local F = {}

local function send_if(sh, name, value)
  if sh:hasUniform(name) then sh:send(name, value) end
end

-- Shadertoy mainImage() wrapped as a LÖVE pixel effect. Replaces texture()
-- with Texel() and injects iTime / iResolution / iChannel0.
function F.convert_shadertoy(src)
  assert(type(src) == "string" and src:find("mainImage"),
    "ellua fx: shadertoy source must define mainImage")
  local body = src
    :gsub("texture2D%s*%(", "Texel(")
    :gsub("texture%s*%(", "Texel(")
  if F.preprocess then body = F.preprocess(body) end
  return table.concat({
    "uniform float iTime;",
    "uniform vec3 iResolution;",
    "uniform Image iChannel0;",
    body,
    [[
    vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
      vec4 fragColor = vec4(0.0);
      mainImage(fragColor, sc);
      return fragColor * color;
    }
    ]],
  }, "\n")
end

local shaders = {}

local function shader(name, src)
  if not shaders[name] then
    shaders[name] = love.graphics.newShader(src)
  end
  return shaders[name]
end

local BRIGHT = [[
  uniform float threshold;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 c = Texel(tex, uv);
    float l = dot(c.rgb, vec3(0.299, 0.587, 0.114));
    float k = smoothstep(threshold, threshold + 0.25, l);
    return vec4(c.rgb * k, c.a) * color;
  }
]]

local BLUR = [[
  uniform vec2 dir;
  vec4 effect(vec4 c, Image tex, vec2 uv, vec2 sc) {
    float w[9];
    w[0]=0.0625; w[1]=0.0938; w[2]=0.1250; w[3]=0.1562; w[4]=0.1688;
    w[5]=0.1562; w[6]=0.1250; w[7]=0.0938; w[8]=0.0625;
    vec4 sum = vec4(0.0);
    float tot = 0.0;
    for (int i = 0; i < 9; i++) {
      float o = float(i) - 4.0;
      sum += Texel(tex, uv + dir * o) * w[i];
      tot += w[i];
    }
    return (sum / tot) * c;
  }
]]

local COMBINE = [[
  uniform Image bloom;
  uniform float strength;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 base = Texel(tex, uv);
    vec4 b = Texel(bloom, uv);
    return vec4(base.rgb + b.rgb * strength, base.a) * color;
  }
]]

local VIGNETTE = [[
  uniform float opacity;
  uniform float radius;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 c = Texel(tex, uv);
    float d = length(uv - vec2(0.5));
    float dark = smoothstep(radius, 1.0, d * 1.45);
    c.rgb *= 1.0 - opacity * dark;
    return c * color;
  }
]]

local CHROMA = [[
  uniform float amount;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec2 dir = (uv - vec2(0.5)) * amount * 0.004;
    float r = Texel(tex, uv + dir).r;
    float g = Texel(tex, uv).g;
    float b = Texel(tex, uv - dir).b;
    float a = Texel(tex, uv).a;
    return vec4(r, g, b, a) * color;
  }
]]

local GRAIN = [[
  uniform float amount;
  uniform float iTime;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 c = Texel(tex, uv);
    float n = fract(sin(dot(sc + iTime * 19.0, vec2(12.9898, 78.233))) * 43758.5453);
    c.rgb += (n - 0.5) * amount;
    return c * color;
  }
]]

local GLOW = [[
  uniform Image glow;
  uniform float strength;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 base = Texel(tex, uv);
    vec4 g = Texel(glow, uv);
    return vec4(max(base.rgb, g.rgb * strength), base.a) * color;
  }
]]

local PASSTHRU = [[
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    return Texel(tex, uv) * color;
  }
]]

-- LYGIA-style snippet table. LÖVE has no preprocessor; concat #include names.
local GLSL = {}
GLSL["lygia/color/tonemap/aces.glsl"] = [[
vec3 aces(vec3 x) {
  const float a = 2.51, b = 0.03, c = 2.43, d = 0.59, e = 0.14;
  return clamp((x * (a * x + b)) / (x * (c * x + d) + e), 0.0, 1.0);
}
]]
GLSL["lygia/filter/kawase/down.glsl"] = [[
vec4 kawaseDown(Image tex, vec2 uv, vec2 pixel) {
  vec4 sum = Texel(tex, uv) * 4.0;
  sum += Texel(tex, uv + vec2(-pixel.x, -pixel.y));
  sum += Texel(tex, uv + vec2( pixel.x, -pixel.y));
  sum += Texel(tex, uv + vec2(-pixel.x,  pixel.y));
  sum += Texel(tex, uv + vec2( pixel.x,  pixel.y));
  return sum / 8.0;
}
]]
GLSL["lygia/filter/kawase/up.glsl"] = [[
vec4 kawaseUp(Image tex, vec2 uv, vec2 pixel) {
  vec2 o = pixel;
  vec4 sum = Texel(tex, uv + vec2(-o.x * 2.0, 0.0));
  sum += Texel(tex, uv + vec2(-o.x, -o.y)) * 2.0;
  sum += Texel(tex, uv + vec2(0.0, -o.y * 2.0));
  sum += Texel(tex, uv + vec2(o.x, -o.y)) * 2.0;
  sum += Texel(tex, uv + vec2(o.x * 2.0, 0.0));
  sum += Texel(tex, uv + vec2(o.x, o.y)) * 2.0;
  sum += Texel(tex, uv + vec2(0.0, o.y * 2.0));
  sum += Texel(tex, uv + vec2(-o.x, o.y)) * 2.0;
  return sum / 12.0;
}
]]
GLSL["lygia/generative/worley.glsl"] = [[
float worley(vec2 uv, float t) {
  vec2 p = uv * 6.0;
  vec2 i = floor(p);
  vec2 f = fract(p);
  float d = 1.0;
  for (int y = -1; y <= 1; y++) {
    for (int x = -1; x <= 1; x++) {
      vec2 g = vec2(float(x), float(y));
      vec2 o = vec2(
        fract(sin(dot(i + g, vec2(127.1, 311.7)) + t) * 43758.5453),
        fract(sin(dot(i + g, vec2(269.5, 183.3)) + t * 1.3) * 43758.5453));
      vec2 r = g + o - f;
      d = min(d, dot(r, r));
    }
  }
  return sqrt(d);
}
]]

function F.preprocess(src)
  local seen = {}
  local function expand(s)
    return (s:gsub('#include%s+"([^"]+)"', function(path)
      if seen[path] then return "" end
      local body = GLSL[path]
      if not body then error("ellua fx: missing include " .. path, 0) end
      seen[path] = true
      return expand(body)
    end))
  end
  return expand(src)
end

local KAWASE_DOWN = F.preprocess([[
  #include "lygia/filter/kawase/down.glsl"
  uniform vec2 pixel;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    return kawaseDown(tex, uv, pixel) * color;
  }
]])

local KAWASE_UP = F.preprocess([[
  #include "lygia/filter/kawase/up.glsl"
  uniform vec2 pixel;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    return kawaseUp(tex, uv, pixel) * color;
  }
]])

local TONEMAP = F.preprocess([[
  #include "lygia/color/tonemap/aces.glsl"
  uniform float amount;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 c = Texel(tex, uv);
    c.rgb = mix(c.rgb, aces(c.rgb * 1.12), amount);
    return c * color;
  }
]])

local PIXELATE = [[
  uniform float amount;
  uniform vec2 pixel;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    float n = mix(1.0, 48.0, amount);
    vec2 cell = pixel * n;
    vec2 uv2 = floor(uv / cell) * cell + cell * 0.5;
    return Texel(tex, uv2) * color;
  }
]]

local POSTERIZE = [[
  uniform float amount;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 c = Texel(tex, uv);
    float levels = mix(12.0, 3.0, amount);
    c.rgb = floor(c.rgb * levels + 0.5) / levels;
    return c * color;
  }
]]

local FILMGRAIN = [[
  uniform float amount;
  uniform float iTime;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 c = Texel(tex, uv);
    float n = fract(sin(dot(sc + iTime * 13.0, vec2(12.9898, 78.233))) * 43758.5453);
    float y = dot(c.rgb, vec3(0.299, 0.587, 0.114));
    c.rgb += (n - 0.5) * amount * (0.55 + 0.45 * y);
    return c * color;
  }
]]

local WORLEY = F.preprocess([[
  #include "lygia/generative/worley.glsl"
  uniform float amount;
  uniform float iTime;
  vec4 effect(vec4 color, Image tex, vec2 uv, vec2 sc) {
    vec4 c = Texel(tex, uv);
    float w = worley(uv + vec2(iTime * 0.03, 0.0), iTime);
    c.rgb = mix(c.rgb, c.rgb * vec3(0.85 + w), amount);
    return c * color;
  }
]])

function F.canvases(node, w, h)
  if node.fx_a and node.fx_w == w and node.fx_h == h then
    return node.fx_a, node.fx_b, node.fx_c
  end
  if node.fx_a then node.fx_a:release() end
  if node.fx_b then node.fx_b:release() end
  if node.fx_c then node.fx_c:release() end
  node.fx_a = love.graphics.newCanvas(w, h)
  node.fx_b = love.graphics.newCanvas(w, h)
  node.fx_c = love.graphics.newCanvas(w, h)
  node.fx_w, node.fx_h = w, h
  return node.fx_a, node.fx_b, node.fx_c
end

local function blit(src, dst, sh)
  local prev = love.graphics.getCanvas()
  local dw, dh = dst:getDimensions()
  local sw, shw = src:getDimensions()
  love.graphics.push("all")
  love.graphics.origin()
  love.graphics.setBlendMode("alpha", "premultiplied")
  love.graphics.setCanvas(dst)
  love.graphics.clear(0, 0, 0, 0)
  love.graphics.setShader(sh)
  love.graphics.setColor(1, 1, 1, 1)
  love.graphics.draw(src, 0, 0, 0, dw / sw, dh / shw)
  love.graphics.setShader()
  love.graphics.setCanvas(prev)
  love.graphics.pop()
end

local function blur_into(src, ping, pong, w, h, radius)
  local sh = shader("blur", BLUR)
  local px = (radius or 6) / w
  local py = (radius or 6) / h
  sh:send("dir", { px, 0 })
  blit(src, ping, sh)
  sh:send("dir", { 0, py })
  blit(ping, pong, sh)
  return pong
end

local function pass_bloom(src, ping, pong, w, h, node)
  local bright = shader("bright", BRIGHT)
  send_if(bright, "threshold", 0.35)
  blit(src, ping, bright)
  local blurred = blur_into(ping, pong, ping, w, h, 8)
  local comb = shader("combine", COMBINE)
  send_if(comb, "bloom", blurred)
  send_if(comb, "strength", node:get("fx_bloom") or 0.45)
  blit(src, pong, comb)
  return pong
end

local function pass_glow(src, ping, pong, w, h, node)
  local blurred = blur_into(src, ping, pong, w, h, 10)
  local sh = shader("glow", GLOW)
  send_if(sh, "glow", blurred)
  send_if(sh, "strength", node:get("fx_glow") or 0.4)
  local dest = (blurred == pong) and ping or pong
  blit(src, dest, sh)
  return dest
end

local function pass_vignette(src, dst, node)
  local sh = shader("vignette", VIGNETTE)
  send_if(sh, "opacity", node:get("fx_vignette") or 0.35)
  send_if(sh, "radius", 0.55)
  blit(src, dst, sh)
  return dst
end

local function pass_chroma(src, dst, node)
  local sh = shader("chroma", CHROMA)
  send_if(sh, "amount", node:get("fx_chroma") or 1.5)
  blit(src, dst, sh)
  return dst
end

local function pass_grain(src, dst, node, t)
  local sh = shader("grain", GRAIN)
  send_if(sh, "amount", node:get("fx_grain") or 0.08)
  send_if(sh, "iTime", t)
  blit(src, dst, sh)
  return dst
end

local function pass_blur(src, ping, pong, w, h, node)
  return blur_into(src, ping, pong, w, h, node:get("fx_blur") or 4)
end

local function pass_shadertoy(src, dst, node, t, w, h)
  local src_code = node.initial.shadertoy
  if not src_code then return src end
  if not node.fx_toy then
    node.fx_toy = love.graphics.newShader(F.convert_shadertoy(src_code))
  end
  local sh = node.fx_toy
  send_if(sh, "iTime", t)
  send_if(sh, "iResolution", { w, h, 1 })
  send_if(sh, "iChannel0", src)
  blit(src, dst, sh)
  return dst
end

local function kawase_lo(node, w, h)
  local hw, hh = math.max(1, math.floor(w / 2)), math.max(1, math.floor(h / 2))
  if node.fx_klo and node.fx_kw == w and node.fx_kh == h then return node.fx_klo end
  if node.fx_klo then node.fx_klo:release() end
  node.fx_klo = love.graphics.newCanvas(hw, hh)
  node.fx_kw, node.fx_kh = w, h
  return node.fx_klo
end

local function pass_kawase(src, dst, node, w, h)
  local lo = kawase_lo(node, w, h)
  local down = shader("kawase_down", KAWASE_DOWN)
  local up = shader("kawase_up", KAWASE_UP)
  local iters = math.max(1, math.min(4, math.floor((node:get("fx_blur") or 2) + 0.5)))
  local cur = src
  for i = 1, iters do
    local px = (0.5 + (i - 1)) / w
    local py = (0.5 + (i - 1)) / h
    down:send("pixel", { px, py })
    blit(cur, lo, down)
    up:send("pixel", { px * 0.5, py * 0.5 })
    blit(lo, dst, up)
    cur = dst
  end
  return dst
end

local function pass_tonemap(src, dst, node)
  local sh = shader("tonemap", TONEMAP)
  send_if(sh, "amount", node:get("fx_tonemap") or 1)
  blit(src, dst, sh)
  return dst
end

local function pass_pixelate(src, dst, node, w, h)
  local sh = shader("pixelate", PIXELATE)
  send_if(sh, "amount", node:get("fx_pixelate") or 0.45)
  send_if(sh, "pixel", { 1 / w, 1 / h })
  blit(src, dst, sh)
  return dst
end

local function pass_posterize(src, dst, node)
  local sh = shader("posterize", POSTERIZE)
  send_if(sh, "amount", node:get("fx_posterize") or 0.55)
  blit(src, dst, sh)
  return dst
end

local function pass_filmgrain(src, dst, node, t)
  local sh = shader("filmgrain", FILMGRAIN)
  send_if(sh, "amount", node:get("fx_grain") or 0.08)
  send_if(sh, "iTime", t)
  blit(src, dst, sh)
  return dst
end

local function pass_worley(src, dst, node, t)
  local sh = shader("worley", WORLEY)
  send_if(sh, "amount", node:get("fx_worley") or 0.45)
  send_if(sh, "iTime", t)
  blit(src, dst, sh)
  return dst
end

-- Apply chain to canvases. `a` holds the unfiltered scene. Returns the canvas
-- that should be drawn.
function F.apply(node, a, b, c, t, w, h)
  local chain = node.initial.chain or { "bloom", "vignette" }
  local src = a
  local function other(x)
    if x == a then return b end
    if x == b then return c end
    return a
  end
  local function pair(x)
    local d = other(x)
    local e = other(d)
    if e == x then e = other(e) end
    return d, e
  end
  for _, name in ipairs(chain) do
    local dst, spare = pair(src)
    if name == "bloom" then
      src = pass_bloom(src, dst, spare, w, h, node)
    elseif name == "glow" then
      src = pass_glow(src, dst, spare, w, h, node)
    elseif name == "blur" then
      src = pass_blur(src, dst, spare, w, h, node)
    elseif name == "vignette" then
      src = pass_vignette(src, dst, node)
    elseif name == "chroma" or name == "chromasep" then
      src = pass_chroma(src, dst, node)
    elseif name == "grain" then
      src = pass_grain(src, dst, node, t)
    elseif name == "shadertoy" then
      src = pass_shadertoy(src, dst, node, t, w, h)
    elseif name == "kawase" then
      src = pass_kawase(src, dst, node, w, h)
    elseif name == "tonemap" or name == "aces" then
      src = pass_tonemap(src, dst, node)
    elseif name == "pixelate" then
      src = pass_pixelate(src, dst, node, w, h)
    elseif name == "posterize" then
      src = pass_posterize(src, dst, node)
    elseif name == "filmgrain" then
      src = pass_filmgrain(src, dst, node, t)
    elseif name == "worley" then
      src = pass_worley(src, dst, node, t)
    else
      error("ellua fx: unknown pass " .. tostring(name), 0)
    end
  end
  return src
end

return F

File: scene/src/fx.rs
//! The `s:fx` chain on the CPU: the same maths as the GLSL passes in
//! `runtime/fx.lua`, run over a premultiplied RGBA8 scratch frame.
//! Pass kinds (opcode 112): 1 bloom, 2 glow, 3 blur, 4 vignette, 5 chroma,
//! 6 grain, 7 tonemap, 8 pixelate, 9 posterize, 10 filmgrain, 11 kawase.
//! worley and shadertoy stay GLSL (escape hatch, lint marks them opaque).

type Px = [f32; 4];

struct Buf {
    w: usize,
    h: usize,
    d: Vec<Px>,
}

impl Buf {
    fn new(w: usize, h: usize) -> Self { Buf { w, h, d: vec![[0.0; 4]; w * h] } }
    fn from_u8(px: &[u8], w: usize, h: usize) -> Self {
        let mut b = Buf::new(w, h);
        for (i, p) in b.d.iter_mut().enumerate() {
            for c in 0..4 { p[c] = px[i * 4 + c] as f32 / 255.0; }
        }
        b
    }
    fn to_u8(&self, px: &mut [u8]) {
        for (i, p) in self.d.iter().enumerate() {
            for c in 0..4 { px[i * 4 + c] = (p[c].clamp(0.0, 1.0) * 255.0 + 0.5) as u8; }
        }
    }
    #[inline]
    fn at(&self, x: i64, y: i64) -> Px {
        // clamp-to-edge, like a LÖVE canvas
        let x = x.clamp(0, self.w as i64 - 1) as usize;
        let y = y.clamp(0, self.h as i64 - 1) as usize;
        self.d[y * self.w + x]
    }
    /// bilinear sample at uv (texel centres at (i+0.5)/w), like `Texel(tex, uv)`
    fn uv(&self, u: f32, v: f32) -> Px {
        let fx = u * self.w as f32 - 0.5;
        let fy = v * self.h as f32 - 0.5;
        let x0 = fx.floor();
        let y0 = fy.floor();
        let tx = fx - x0;
        let ty = fy - y0;
        let (x0, y0) = (x0 as i64, y0 as i64);
        let a = self.at(x0, y0);
        let b = self.at(x0 + 1, y0);
        let c = self.at(x0, y0 + 1);
        let d = self.at(x0 + 1, y0 + 1);
        let mut o = [0.0; 4];
        for i in 0..4 {
            o[i] = (a[i] * (1.0 - tx) + b[i] * tx) * (1.0 - ty) + (c[i] * (1.0 - tx) + d[i] * tx) * ty;
        }
        o
    }
    fn map(&self, f: impl Fn(usize, usize, Px) -> Px + Sync) -> Buf {
        let mut o = Buf::new(self.w, self.h);
        for y in 0..self.h {
            for x in 0..self.w {
                o.d[y * self.w + x] = f(x, y, self.d[y * self.w + x]);
            }
        }
        o
    }
}

#[inline]
fn smoothstep(e0: f32, e1: f32, x: f32) -> f32 {
    let t = ((x - e0) / (e1 - e0)).clamp(0.0, 1.0);
    t * t * (3.0 - 2.0 * t)
}
#[inline]
fn luma(p: Px) -> f32 { p[0] * 0.299 + p[1] * 0.587 + p[2] * 0.114 }
#[inline]
fn fract(x: f32) -> f32 { x - x.floor() }
#[inline]
fn mix(a: f32, b: f32, t: f32) -> f32 { a + (b - a) * t }

const BLUR_W: [f32; 9] = [0.0625, 0.0938, 0.1250, 0.1562, 0.1688, 0.1562, 0.1250, 0.0938, 0.0625];

/// 9-tap separable blur with taps `radius` px apart (runtime/fx.lua BLUR)
fn blur(src: &Buf, radius: f32) -> Buf {
    let tot: f32 = BLUR_W.iter().sum();
    let pass = |b: &Buf, dx: f32, dy: f32| -> Buf {
        b.map(|x, y, _| {
            let (u, v) = ((x as f32 + 0.5) / b.w as f32, (y as f32 + 0.5) / b.h as f32);
            let mut s = [0.0f32; 4];
            for (i, w) in BLUR_W.iter().enumerate() {
                let o = i as f32 - 4.0;
                let p = b.uv(u + dx * o, v + dy * o);
                for c in 0..4 { s[c] += p[c] * w; }
            }
            for c in 0..4 { s[c] /= tot; }
            s
        })
    };
    let ping = pass(src, radius / src.w as f32, 0.0);
    pass(&ping, 0.0, radius / src.h as f32)
}

fn bloom(src: &Buf, strength: f32) -> Buf {
    let bright = src.map(|_, _, c| {
        let k = smoothstep(0.35, 0.60, luma(c));
        [c[0] * k, c[1] * k, c[2] * k, c[3]]
    });
    let b = blur(&bright, 8.0);
    src.map(|x, y, c| {
        let g = b.d[y * b.w + x];
        [c[0] + g[0] * strength, c[1] + g[1] * strength, c[2] + g[2] * strength, c[3]]
    })
}

fn glow(src: &Buf, strength: f32) -> Buf {
    let b = blur(src, 10.0);
    src.map(|x, y, c| {
        let g = b.d[y * b.w + x];
        [c[0].max(g[0] * strength), c[1].max(g[1] * strength), c[2].max(g[2] * strength), c[3]]
    })
}

fn vignette(src: &Buf, opacity: f32) -> Buf {
    src.map(|x, y, c| {
        let (u, v) = ((x as f32 + 0.5) / src.w as f32 - 0.5, (y as f32 + 0.5) / src.h as f32 - 0.5);
        let d = (u * u + v * v).sqrt();
        let dark = smoothstep(0.55, 1.0, d * 1.45);
        let k = 1.0 - opacity * dark;
        [c[0] * k, c[1] * k, c[2] * k, c[3]]
    })
}

fn chroma(src: &Buf, amount: f32) -> Buf {
    src.map(|x, y, c| {
        let (u, v) = ((x as f32 + 0.5) / src.w as f32, (y as f32 + 0.5) / src.h as f32);
        let (dx, dy) = ((u - 0.5) * amount * 0.004, (v - 0.5) * amount * 0.004);
        let r = src.uv(u + dx, v + dy)[0];
        let b = src.uv(u - dx, v - dy)[2];
        [r, c[1], b, c[3]]
    })
}

fn noise(x: f32, y: f32, t: f32, rate: f32) -> f32 {
    let sx = x + 0.5 + t * rate;
    let sy = y + 0.5 + t * rate;
    fract((sx * 12.9898 + sy * 78.233).sin() * 43758.5453)
}

fn grain(src: &Buf, amount: f32, t: f32) -> Buf {
    src.map(|x, y, c| {
        let n = noise(x as f32, y as f32, t, 19.0) - 0.5;
        [c[0] + n * amount, c[1] + n * amount, c[2] + n * amount, c[3]]
    })
}

fn filmgrain(src: &Buf, amount: f32, t: f32) -> Buf {
    src.map(|x, y, c| {
        let n = (noise(x as f32, y as f32, t, 13.0) - 0.5) * amount * (0.55 + 0.45 * luma(c));
        [c[0] + n, c[1] + n, c[2] + n, c[3]]
    })
}

fn aces(x: f32) -> f32 {
    let (a, b, c, d, e) = (2.51, 0.03, 2.43, 0.59, 0.14);
    ((x * (a * x + b)) / (x * (c * x + d) + e)).clamp(0.0, 1.0)
}

fn tonemap(src: &Buf, amount: f32) -> Buf {
    src.map(|_, _, c| [
        mix(c[0], aces(c[0] * 1.12), amount),
        mix(c[1], aces(c[1] * 1.12), amount),
        mix(c[2], aces(c[2] * 1.12), amount),
        c[3],
    ])
}

fn pixelate(src: &Buf, amount: f32) -> Buf {
    let n = mix(1.0, 48.0, amount);
    src.map(|x, y, _| {
        let (u, v) = ((x as f32 + 0.5) / src.w as f32, (y as f32 + 0.5) / src.h as f32);
        let (cu, cv) = (n / src.w as f32, n / src.h as f32);
        src.uv((u / cu).floor() * cu + cu * 0.5, (v / cv).floor() * cv + cv * 0.5)
    })
}

fn posterize(src: &Buf, amount: f32) -> Buf {
    let levels = mix(12.0, 3.0, amount);
    src.map(|_, _, c| [
        (c[0] * levels + 0.5).floor() / levels,
        (c[1] * levels + 0.5).floor() / levels,
        (c[2] * levels + 0.5).floor() / levels,
        c[3],
    ])
}

fn kawase(src: &Buf, iters: f32) -> Buf {
    let iters = (iters + 0.5).floor().clamp(1.0, 4.0) as usize;
    let (hw, hh) = ((src.w / 2).max(1), (src.h / 2).max(1));
    let mut cur = Buf { w: src.w, h: src.h, d: src.d.clone() };
    for i in 1..=iters {
        let (px, py) = ((0.5 + (i - 1) as f32) / src.w as f32, (0.5 + (i - 1) as f32) / src.h as f32);
        let lo = Buf::new(hw, hh).map(|x, y, _| {
            let (u, v) = ((x as f32 + 0.5) / hw as f32, (y as f32 + 0.5) / hh as f32);
            let mut s = cur.uv(u, v);
            for c in 0..4 { s[c] *= 4.0; }
            for (ox, oy) in [(-px, -py), (px, -py), (-px, py), (px, py)] {
                let p = cur.uv(u + ox, v + oy);
                for c in 0..4 { s[c] += p[c]; }
            }
            for c in 0..4 { s[c] /= 8.0; }
            s
        });
        let (ox, oy) = (px * 0.5, py * 0.5);
        cur = Buf::new(src.w, src.h).map(|x, y, _| {
            let (u, v) = ((x as f32 + 0.5) / src.w as f32, (y as f32 + 0.5) / src.h as f32);
            let taps: [(f32, f32, f32); 8] = [
                (-ox * 2.0, 0.0, 1.0), (-ox, -oy, 2.0), (0.0, -oy * 2.0, 1.0), (ox, -oy, 2.0),
                (ox * 2.0, 0.0, 1.0), (ox, oy, 2.0), (0.0, oy * 2.0, 1.0), (-ox, oy, 2.0),
            ];
            let mut s = [0.0f32; 4];
            for (tx, ty, w) in taps {
                let p = lo.uv(u + tx, v + ty);
                for c in 0..4 { s[c] += p[c] * w; }
            }
            for c in 0..4 { s[c] /= 12.0; }
            s
        });
    }
    cur
}

/// Run `passes` (kind, amount, extra) over a premultiplied RGBA8 frame in place.
pub fn run_chain(px: &mut [u8], w: usize, h: usize, passes: &[(u32, f32, f32)]) {
    if w == 0 || h == 0 || passes.is_empty() { return; }
    let mut cur = Buf::from_u8(px, w, h);
    for &(kind, amount, extra) in passes {
        cur = match kind {
            1 => bloom(&cur, amount),
            2 => glow(&cur, amount),
            3 => blur(&cur, amount),
            4 => vignette(&cur, amount),
            5 => chroma(&cur, amount),
            6 => grain(&cur, amount, extra),
            7 => tonemap(&cur, amount),
            8 => pixelate(&cur, amount),
            9 => posterize(&cur, amount),
            10 => filmgrain(&cur, amount, extra),
            11 => kawase(&cur, amount),
            _ => cur,
        };
    }
    cur.to_u8(px);
}
